src/main/java/com/microsoft/azure/vmagent/AzureVMAgentCleanUpTask.java

```diff
@@ -185,7 +201,121 @@ else if(state.equalsIgnoreCase("succeeded") && diffTimeInMinutes > successTimeou
+        List<String> VMs = new ArrayList<>();
+        Jenkins instance = Jenkins.getInstance();
+        if (instance != null) {
+            for (Computer computer : instance.getComputers()) {
+                if (computer instanceof AzureVMComputer) {
+                    AzureVMComputer azureComputer = (AzureVMComputer) computer;
+                    AzureVMAgent agent = azureComputer.getNode();
+                    if (agent != null && agent.getCloudName().equals(cloudName)) {
+                        final String vmName = computer.getName();
+                        VMs.add(vmName);
+                    }
+                }
+            }
+        }
+        return VMs;
+    }
+
+    public void cleanLeakedResources(
+            final String resourceGroup,
+            final ServicePrincipal servicePrincipal,
+            final String cloudName,
+            final DeploymentRegistrar deploymentRegistrar) {
+        try{
+            final List<String> validVMs = getValidVMs(cloudName);
+            final Azure azureClient  = TokenCache.getInstance(servicePrincipal).getAzureClient();
+            //can't use listByTag because for some reason that method strips all the tags from the outputted resources (https://github.com/Azure/azure-sdk-for-java/issues/1436)
+            final PagedList<GenericResource> resources = azureClient.genericResources().listByGroup(resourceGroup);
+
+            final PriorityQueue<GenericResource> resourcesMarkedForDeletion = new PriorityQueue<> (resources.size(), new Comparator<GenericResource>(){
+                @Override
+                public int compare(GenericResource o1, GenericResource o2) {
+                    int o1Priority = getPriority(o1);
+                    int o2Priority = getPriority(o2);
+                    if (o1Priority == o2Priority) {
+                        return 0;
+                    }
+                    return (o1Priority < o2Priority) ? - 1 : 1;
+                }
+                private int getPriority(final GenericResource resource) {
+                    final String type = resource.type();
+                    if (StringUtils.containsIgnoreCase(type, "virtualMachine")) {
+                        return 1;
+                    }
+                    if (StringUtils.containsIgnoreCase(type, "networkInterface")) {
+                        return 2;
+                    }
+                    if (StringUtils.containsIgnoreCase(type, "IPAddress")) {
+                        return 3;
+                    }
+                    return 4;
+                }
+            });
+
+            for (GenericResource resource : resources) {
+                final Map<String,String> tags = resource.tags();
+                if ( !tags.containsKey(Constants.AZURE_RESOURCES_TAG_NAME) || 
+                     !deploymentRegistrar.getDeploymentTag().matches(new AzureUtil.DeploymentTag(tags.get(Constants.AZURE_RESOURCES_TAG_NAME)))) {
```

Comment:
It's just a cleaner way to compare the timestamps and the Jenkins instance id. I just wrap the tag and compare it against whatever the current tag would be.